# P-value for feature reduction

In [1]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)
import warnings
warnings.filterwarnings('ignore')
import statsmodels.api as sm
from sklearn.datasets import load_boston

## Fitting OLS model

In [2]:
x = load_boston() # Boston home prices dataset from scikit learn
y = x.target # price
X = pd.DataFrame(x.data, columns = x.feature_names)

X_added_constant = sm.add_constant(X) # we need to add this constant for the intercept w/ OLS
X_added_constant # input features for linear regression use OLS

const     CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  \
0      1.0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0   
1      1.0  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0   
2      1.0  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0   
3      1.0  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0   
4      1.0  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0   
..     ...      ...   ...    ...   ...    ...    ...   ...     ...  ...   
501    1.0  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786  1.0   
502    1.0  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875  1.0   
503    1.0  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675  1.0   
504    1.0  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889  1.0   
505    1.0  0.04741   0.0  11.93   0.0  0.573  6.030  80.8  2.5050  1.0   

       TAX  PTRATIO       B  LSTAT  
0    296.0     15.3  396.90   4.98  
1    242.0     17.8  396.90   9.14  
2    242.0     17.8  392.83   4.03  
3    222.0     18.7  394.63   2.94  
4    222.0     18.7  396.90   5.33  
..     ...      ...     ...    ...  
501  273.0     21.0  391.99   9.67  
502  273.0     21.0  396.90   9.08  
503  273.0     21.0  396.90   5.64  
504  273.0     21.0  393.45   6.48  
505  273.0     21.0  396.90   7.88  

[506 rows x 14 columns]

In [3]:
model = sm.OLS(y,X_added_constant).fit() # ordinary least squares regression model from statsmodels
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.741
Model:                            OLS   Adj. R-squared:                  0.734
Method:                 Least Squares   F-statistic:                     108.1
Date:                Fri, 09 Dec 2022   Prob (F-statistic):          6.72e-135
Time:                        00:32:42   Log-Likelihood:                -1498.8
No. Observations:                 506   AIC:                             3026.
Df Residuals:                     492   BIC:                             3085.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         36.4595      5.103      7.144      0.000      26.432      46.487
CRIM          -0.1080      0.033     -3.287      0.001      -0.173      -0.043
ZN             0.0464      0.014      3.382      0.001       0.019       0.073
INDUS          0.0206      0.061      0.334      0.738      -0.100       0.141
CHAS           2.6867      0.862      3.118      0.002       0.994       4.380
NOX          -17.7666      3.820     -4.651      0.000     -25.272     -10.262
RM             3.8099      0.418      9.116      0.000       2.989       4.631
AGE            0.0007      0.013      0.052      0.958      -0.025       0.027
DIS           -1.4756      0.199     -7.398      0.000      -1.867      -1.084
RAD            0.3060      0.066      4.613      0.000       0.176       0.436
TAX           -0.0123      0.004     -3.280      0.001      -0.020      -0.005
PTRATIO       -0.9527      0.131     -7.283      0.000      -1.210      -0.696
B              0.0093      0.003      3.467      0.001       0.004       0.015
LSTAT         -0.5248      0.051    -10.347      0.000      -0.624      -0.425
==============================================================================
Omnibus:                      178.041   Durbin-Watson:                   1.078
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              783.126
Skew:                           1.521   Prob(JB):                    8.84e-171
Kurtosis:                       8.281   Cond. No.                     1.51e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.51e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

## Dropping Insignificant Features using p-value thresholding

In practice, p-value is used for feature selection in the following way: we set the significance level alpha to 0.05, so that when p<0.05, we reject the null hypothesis and consider the feature to be important. p-value in this context corresponds to the probability that a particular feature doesn't affect the model (null hypothesis is true).

['const',
 'CRIM',
 'ZN',
 'CHAS',
 'NOX',
 'RM',
 'DIS',
 'RAD',
 'TAX',
 'PTRATIO',
 'B',
 'LSTAT']

In [9]:
alpha                = 0.05 # significance threshold
sig_param_names      = model.params[list(np.where(model.pvalues < alpha)[0])].iloc[0:].index.tolist() # list of names of signficant parameters
significant_features = X_added_constant[sig_param_names] # only select vars having p-value < 0.05
significant_features # print out the significant feature values

const     CRIM    ZN  CHAS    NOX     RM     DIS  RAD    TAX  PTRATIO  \
0      1.0  0.00632  18.0   0.0  0.538  6.575  4.0900  1.0  296.0     15.3   
1      1.0  0.02731   0.0   0.0  0.469  6.421  4.9671  2.0  242.0     17.8   
2      1.0  0.02729   0.0   0.0  0.469  7.185  4.9671  2.0  242.0     17.8   
3      1.0  0.03237   0.0   0.0  0.458  6.998  6.0622  3.0  222.0     18.7   
4      1.0  0.06905   0.0   0.0  0.458  7.147  6.0622  3.0  222.0     18.7   
..     ...      ...   ...   ...    ...    ...     ...  ...    ...      ...   
501    1.0  0.06263   0.0   0.0  0.573  6.593  2.4786  1.0  273.0     21.0   
502    1.0  0.04527   0.0   0.0  0.573  6.120  2.2875  1.0  273.0     21.0   
503    1.0  0.06076   0.0   0.0  0.573  6.976  2.1675  1.0  273.0     21.0   
504    1.0  0.10959   0.0   0.0  0.573  6.794  2.3889  1.0  273.0     21.0   
505    1.0  0.04741   0.0   0.0  0.573  6.030  2.5050  1.0  273.0     21.0   

          B  LSTAT  
0    396.90   4.98  
1    396.90   9.14  
2    392.83   4.03  
3    394.63   2.94  
4    396.90   5.33  
..      ...    ...  
501  391.99   9.67  
502  396.90   9.08  
503  396.90   5.64  
504  393.45   6.48  
505  396.90   7.88  

[506 rows x 12 columns]

In [10]:
# rerun the model with only the features having p<0.05
model            = sm.OLS(y,X_added_constant).fit()
model.summary()

KeyError: "['INDUS', 'AGE'] not found in axis"

<b> We see that after dropping one insignificant variable, the adjusted R-squared improved slightly (from 0.734
to 0.735)